Change how new competitions are created - fixes #416

pingskills opened the pull request:
The new page and edit page now share a partial form. The id and cellName are only validated if the name passes validation or it's an update. This is so the error messages can be simplified. Competitions can now be cloned from the competition nav bar.

Let me know if you think this is the right direction to head in. I think it simplifies the code and the user process flow. The clone competition functionality has now moved from the model to the controller.

Here are some screenshots:
The new page for creating a competition:
![create-comp-new-screen](https://cloud.githubusercontent.com/assets/4403168/15447569/f477077e-1f89-11e6-9d6b-4cadaf7fc276.png)

The error message if you don't add a year to the name:
![create-comp-error-messages](https://cloud.githubusercontent.com/assets/4403168/15447570/fec9052e-1f89-11e6-97a0-0785ef00a151.png)

The navigation bar with the clone link:
![create-comp-nav-bar](https://cloud.githubusercontent.com/assets/4403168/15447571/058e27cc-1f8a-11e6-8566-3d9201b27e1e.png)

The new screen after the clone link is hit:
![create-comp-clone](https://cloud.githubusercontent.com/assets/4403168/15447575/4e3925ee-1f8a-11e6-9fb6-f56202aab5dd.png)


I'd love to hear your thoughts

jonatanklosko commented:
Now it's not obvious that you can clone the competition so I'd see here an info alert on the new competition page informing about this.

jonatanklosko commented on WcaOnRails/app/assets/javascripts/competitions.js:
```diff
@@ -1,26 +1,4 @@
-});
-
-onPage('competitions#edit, competitions#update', function() {
+onPage('competitions#edit, competitions#new', function() {
```
You have to specify all the *actions* that renders the new/edit pages:
`'competitions#new competitions#create competitions#edit competitions#admin_edit competitions#clone_competition'`

jonatanklosko commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -239,16 +232,35 @@ def edit
+    @competition.showAtAll = false
+    @competition.isConfirmed = false
+    render :new
+  end
```
I'd like to have this logic inside the model to slim down the controller as much as we can (it's big enough). In the current situation, instead of using the `before_validation` callback as before, it would be better just to make a method. Something like:
```ruby
# Called for the competition which we want to clone data from
def get_clone
  # All the logic here
  # Return a *new* competition with the attributes assigned
end
```

Then in the controller something like this:
```ruby
cloned_competition = Competition.find(params[:id])
@competition = cloned_competition.get_clone
```

jfly replied:
I like this suggestion. Maybe to follow FactoryGirl semantics, we could call the method `build_clone`?

jonatanklosko replied:
Awesome!

pingskills replied:
OK. This makes sense but I think I need to keep the following in the controller since in the model it won't know about curent_user:

`    if current_user.any_kind_of_delegate?
      @clone.delegates |= [current_user]
    end`

jonatanklosko replied:
It would be sufficient to move only this if statement to controller and keep the rest inside the model.

jonatanklosko replied:
Oh sorry I see that's what you've done already =)

jonatanklosko commented on WcaOnRails/app/views/competitions/_nav.html.erb:
```diff
@@ -50,6 +50,13 @@
+        <% nav_items << {
+          text: t('.menu.clone'),
+          path: clone_competition_path(@competition),
+          fa_icon: "lock",
```
There is a ['clone' icon](http://fontawesome.io/icon/clone/). I'm not making any suggestion, just check it out and use whatever you find cooler =)

jonatanklosko commented on WcaOnRails/app/views/competitions/new.html.erb:
```diff
@@ -1,14 +1,7 @@
 <% provide(:title, t('.create_competition')) %>
+<% provide(:include_gmaps, true) %>
```
Does it work if we remove this line both from here and `edit.html` and move it to `_competition_form.html.erb` ?

pingskills replied:
It does work :)

jonatanklosko replied:
Great =)

jonatanklosko commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -239,16 +232,35 @@ def edit
-    @competition = Competition.find(params[:id])
+    @competition = Competition.new
     @competition.assign_attributes(competition_params)
     @competition.valid? # We only unpack dates _just before_ validation, so we need to call validation here
```
I feel like it would be much more expressive to make `unpack_dates` public and call it here instead of this workaround (if there are no obstacles to do so).

pingskills replied:
At the moment as it's called for validation purposes it adds to errors. If changed this to only unpack the dates we'd need to do the validation somewhere else. I'm not sure what is best here?

jonatanklosko replied:
Ok, that just wasn't obvious from the comment that we need to add errors too.

jonatanklosko commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -239,16 +232,35 @@ def edit
-    @competition = Competition.find(params[:id])
+    @competition = Competition.new
     @competition.assign_attributes(competition_params)
     @competition.valid? # We only unpack dates _just before_ validation, so we need to call validation here
```
Same as above.

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -331,7 +311,11 @@ def latitude_radians
   def events
     # See https://github.com/cubing/worldcubeassociation.org/issues/95 for
```
nit: I think it makes more sense for this comment to live next to the line of code `eventSpecs.split...`

jfly commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -239,16 +232,35 @@ def edit
+
   def nearby_competitions
-    @competition = Competition.find(params[:id])
+    @competition = Competition.new
```
Good call.

jfly commented on WcaOnRails/spec/features/create_competition_spec.rb:
```diff
@@ -14,43 +14,29 @@
 
   it 'can clone competition' do
-    visit "/competitions/new"
+    visit clone_competition_path(competition_to_clone)
```
MAN THIS TEST IS SO NICE NOW

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -21,12 +21,15 @@ class Competition < ActiveRecord::Base
-                 format: { with: /\A[a-zA-Z0-9]+\Z/ }
+                 format: { with: /\A[a-zA-Z0-9]+\Z/ }, if: :name_valid_or_updating?
+  private def name_valid_or_updating?
+    self.persisted? ? true : name.length <= MAX_NAME_LENGTH && name =~ ENDS_WITH_YEAR_RE
```
I don't think the ternary is serving any purpose here. Can this be rewritten as follows?

```
self.persisted? || (name.length <= MAX_NAME_LENGTH && name =~ ENDS_WITH_YEAR_RE)
```

More importantly, what is going on here? Why do we need to look at `.persisted?` now?

pingskills replied:
I want to perform this validation all the time on the edit page. Because someone could update the name and the cellName on the validation page and at that point I want to display both errors to the user. It's only on the new page that I don't want to validate the id and cellName if the name itself isn't valid.

jfly commented:
This is some awesome cleanup, @pingskills! I'm very pleased with this new direction.

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -331,7 +311,11 @@ def latitude_radians
+      eventSpecs.split.map { |e| Event.find_by_id(e.split("=")[0]) }.sort_by &:rank
+    else
+      []
+    end
```
`(eventSpecs || []).split.map ...`

pingskills commented:
Thanks for your feedback everyone. I've updated the code to try and reflect your suggestions. Let me know what you think.

pingskills commented:
When smoke testing I went to create a new competition, filled in all the details and tried to Confirm the competition (rather than create then confirm) but I got an error with the following code:

`    if params[:commit] == "Confirm" && current_user.can_confirm_competition?(@competition)
      competition_params[:isConfirmed] = true
    end`

This tries to call 
`  def can_confirm_competition?(competition)
    # We don't let competition organizers confirm competitions.
    can_admin_results? || competition.delegates.include?(self)
  end`

but as the competition only has delegate_ids and not any delegates at this point the competition.delegates.include?(self) fails. Any suggestions on how to handle this? I could make the new page only have a Create button but not a Confirm button?

jonatanklosko commented on WcaOnRails/app/views/competitions/new.html.erb:
```diff
@@ -2,13 +2,6 @@
-
-    <button type="submit" id="create-competition" class="btn btn-success"></button>
-  <% end %>
+  <%= alert :info, 'To clone an existing competition visit the information page for that competition and click the "Clone" link in the menu.', note: true %>
```
Nice, anyway I think it shouldn't appear when we are already cloning a competition. So probably it would be better to set this alert in `flash.now` in the *new* action.

jfly replied:
Or alternatively, you could check if the action is `new` here, rather than `create`.

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -21,12 +21,15 @@ class Competition < ActiveRecord::Base
-                 format: { with: /\A[a-zA-Z0-9]+\Z/ }
+                 format: { with: /\A[a-zA-Z0-9]+\Z/ }, if: :name_valid_or_updating?
+  private def name_valid_or_updating?
+    self.persisted? || name.length <= MAX_NAME_LENGTH && name =~ ENDS_WITH_YEAR_RE
```
So, I'm still confused. Based on the boolean logic, this looks more like `name_valid_or_creating?`, or even `creating_or_name_valid?`.

> I want to perform this validation all the time on the edit page. Because someone could update the name and the cellName on the validation page and at that point I want to display both errors to the user. It's only on the new page that I don't want to validate the id and cellName if the name itself isn't valid.

I see. I'm nervous that this validation would allow people to create new competitions with invalid `id`s and `cellName`s, though. You could achieve this by adding a form manually via javascript, or using curl.

For now, I think it would be okay to just always enforce these validations. Admittedly, it would be confusing to users to get validation errors on fields they don't even see, but I think we could eventually port our automatic `id` and `cellName` computation to javascript, and get rid of the yucky mess inside our model to conditionally compute them. I think this would clean up our code, and provide a much nicer UX.

jfly replied:
PLEASE FORGET EVERYTHING I JUST SAID, @jonatanklosko SET ME STRAIGHT.

This looks great to me.

jonatanklosko replied:
Despite of the fact that `&&` is stronger, I feel like parentheses should go here for explicitness. If you think that's personal feeling and you don't think the same, just ignore this comment =)

pingskills replied:
That makes sense to me @jonatanklosko. I've added the parentheses.

LinusFresz commented:
Hey, 
Awesome work! This looks much more pleasant to create a competition!
I am not totally sure if this is already fixed (because mostly I don't understand what happens here), so I will just leave a quick note.

While creating a new competition today I experienced some trouble in cloning a former competition. I created a new competition for the German Nationals 2016 (officially confirmed yet) and wanted to clone the German Nationals 2015 page, as it takes place at the same venue. I got the following error:

The form contains 1 error:
WCA delegate(s) are not all delegates

I know where this problem comes from, as Björn Korbanka, who was one of the delegates at this competition, is not a delegate anymore. He retired at the end of last year.

Here is my request: would it be possible to be able to clone this competitions with former delegates who are not delegates anymore? I feel like there could be some warning ('WCA delegate(s) are not all delegates’) and in the cloning process they simply get deleted from the delegate list for this competition, but not be able to clone this competitions is kind of annoying, as all other information are the ones needed.

Never mind this post if you already fixed that! :)

All the best,
Linus

larspetrus commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -29,6 +29,9 @@ class CompetitionsController < ApplicationController
   def new
     @competition = Competition.new
+    if current_user.any_kind_of_delegate?
+      @competition.delegates |= [current_user]
```
Is this meant to be `||=` ?

jonatanklosko replied:
Came up with the same thought but it shouldn't. When using `||=` we expect that `@competition.delegates` is nil, whereas it's `#<ActiveRecord::Associations::CollectionProxy []>`. So we need to get the union here.

larspetrus replied:
Aha! Don't think I've ever encountered `Array.|` before!

I'm used to seeing `@competition.delegates << current_user` to do this.

jfly replied:
I agree, in this particular case, `<<` seems better.

jonatanklosko replied:
Agree, with this someone may stumble upon this and have a temptation to change it (as me and Lars =) ).

pingskills replied:
If I clone a competition that I am already a delegate for and use << then I'll end up with my user id twice in the delegate_ids, but it seems that simple_form handles this so maybe it's not an issue?

jfly replied:
Are you sure that happens? I would have expected the fact that we're using ActiveRecord to prevent such a thing.

In this particular case, we just created a competition, so there are no delegates. Perhaps it would be cleanest to do:

```ruby
@competition.delegates = [current_user]
```

instead?

pingskills replied:
Hi @jfly, I think you're right ActiveRecord must take care of this. If I clone a competition I'm already a delegate for I get the following values in delegate_ids, `value="6180,6181,6183,6180"` where I'm user 6180. But when I save the competition only the 3 unique records get stored in the database. So there are no real problems using << 

I think as this is cloning a competition we'd want to also clone the existing delegates rather than just set it to the current user.

pingskills replied:
Actually although it saves the correct records to the database one of the test fails because of the duplicate record. I might leave it as |=

Here's the test that fails:

`      it 'clones a competition that they delegated' do
        # First, make ourselves the delegate of the competition we're going to clone.
        competition.delegates = [delegate]
        get :clone_competition, id: competition
        new_comp = assigns(:competition)
        expect(new_comp.id).to eq ""

        # Cloning a competition should clone its organizers.
        expect(new_comp.organizers.sort_by(&:id)).to eq []
        # When a delegate clones a competition, it should clone its organizers, and add
        # the delegate doing the cloning.
        expect(new_comp.delegates.sort_by(&:id)).to eq [delegate]
      end`

It's because the delegate is there twice.

larspetrus replied:
I tried looking into this stuff, and it's fairly contentious and confusing. It seems Rails does this right with direct associations, if you specify :uniq on the relation, but not on `through` relations, and that's (debatably) a bug.

Or something... This feels like a bottomless hole so I stopped digging.

https://github.com/rails/rails/issues/8573

pingskills commented:
Does anyone have a suggestion on how to deal with the following code called when confirming a comp? 

`current_user.can_confirm_competition?(@competition)`

I have the confirm button on the new page, and this method checks competition.delegates which doesn't exist for a new competition? We only have the delegate_ids in the parameters.

I could take the confirm button off the new page, meaning a user would need to first create, and then they could confirm but it seems like an unnecessary step. I'd rather change the can_confirm_competition to check if @competition.delegates is nil and then check the delegate_ids passed in the parameters.

pingskills commented:
I modified the controller where it checks if a user can confirm a competition to first see if we have a @competition, otherwise check the delegate_ids in the params.

pingskills commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -359,8 +361,15 @@ def my_competitions
+        if current_user.can_confirm_competition?(@competition)
+          competition_params[:isConfirmed] = true
+        end
+      elsif competition_params[:delegate_ids].split(",").include? current_user.id.to_s
```
If we confirm a competition from the new page there is no @competition at this point in the code so in that case I'm checking that the user is in the delegate_ids to ensure they can confirm this competition.

jfly replied:
Yikes. I'd rather remove the ability to confirm a competition from the New page. No delegate in their right mind is going to fill out that entire form without wanting to hit save a few times before confirming.

pingskills replied:
That will certainly make it simpler. I'll revert this and remove the confirm button in this situation.

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -96,25 +99,17 @@ def info
+    clone.delegates = delegates
+    clone.showAtAll = false
+    clone.isConfirmed = false
+    clone
```
I feel like cleaning this up. What about:
```ruby
NOT_CLONED_ATTRIBUTES = %w(id name cellName year month day endMonth endDay registration_open registration_close results_posted_at)
Competition.new(attributes.except(*NOT_CLONED_ATTRIBUTES)).tap do |clone|
  clone.assign_attributes showAtAll: false, isConfirmed: false
  clone.organizers = organizers
  clone.delegates = delegates
end
```

larspetrus replied:
Yes, that really needs cleaning up!

I'd be a bit more ambitious and define both clonable and unclonable attributes. That makes this test possible...

`expect(Competition.column_names()).to match_array(Competition::CLONABLE_ATTRS + Competition::UNCLONABLE_ATTRS)`

..., which makes the clone code self maintaining! In the sense that if a column is added or renamed, the test will go Red, and we have to explicitly declare how the new column is treated.


Also, `showAtAll` and `isConfirmed` are set to `false` by the schema, so it seems wrong to set it here too.

jonatanklosko replied:
Really nice idea with this explicit division!
Also that's a good note, it would be sufficient to add `showAtAll` and `isConfirmed` to unclonable attributes rather than this assigning =)

jonatanklosko commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -104,20 +107,10 @@ def index
-    new_competition_params = params.require(:competition).permit(:name, :competition_id_to_clone)
-    @competition = Competition.new(new_competition_params.merge(
-                                   registration_open: 1.week.from_now,
-                                   registration_close: 2.weeks.from_now))
```
Why have you removed the registration dates from here? Maybe we can handle this in the model with
```ruby
after_initialize :set_default_registration_dates
private def set_default_registration_dates
  registration_open ||= # default
  registration_close ||= # default
```
@larspetrus @jfly do you see any cleaner way?

larspetrus replied:
Depends on why we are/were assigning arbitrary values to these fields.

Without knowing the goal it's hard to say what the best way to reach it is.

jonatanklosko replied:
Oh, actually it doesn't seem to be necessary. I just assumed there was a reason for that. Nevermind!

jfly replied:
I think this was added in 75de58d46e3c474f260a4c83cdbb7afcf3beda3f because `registration_open` and `registration_close` were both `NOT NULL`.

Later on (17506dd8131cb30c6a3c9bc9925f8755cee84eab), it looks like we changed these fields to be nullable, but didn't change this code to no longer set `registration_open` and `registration_close`. I support removing these defaults now.

pingskills commented:
Cleaned up `build_clone` as suggested by @jonatanklosko and added new test as suggested by @larspetrus. I also removed the confirm button from the new page.

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -100,16 +102,10 @@ def info
-    clone.showAtAll = false
-    clone.isConfirmed = false
-    clone
+    Competition.new(attributes.except(*UNCLONEABLE_ATTRIBUTES)).tap do |clone|
```
As you've added both attributes groups as Lars suggested, it would be better to do `attributes.slice(CLONABLE_ATTRIBUTES)` here.

This really cleaned things up.

jonatanklosko commented on WcaOnRails/spec/models/competition_spec.rb:
```diff
@@ -206,6 +206,10 @@
   end
 
+  it "ensures all attributes are defined as either cloneable or uncloneable" do
+    expect(Competition.column_names()).to match_array(Competition::CLONEABLE_ATTRIBUTES + Competition::UNCLONEABLE_ATTRIBUTES)
```
Please remove this empty `()` =)

jonatanklosko commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -96,24 +101,10 @@ def info
-        errors.add(:competition_id_to_clone, "invalid")
-      end
+  def build_clone
+    Competition.new(attributes.except(*UNCLONEABLE_ATTRIBUTES)).tap do |clone|
```
As you've added both attributes groups as Lars suggested, it would be better to do
```ruby
attributes.slice(CLONABLE_ATTRIBUTES)
```

This really cleaned things up.

jonatanklosko commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -239,16 +232,25 @@ def edit
   def nearby_competitions
-    @competition = Competition.find(params[:id])
+    @competition = Competition.new
     @competition.assign_attributes(competition_params)
```
```ruby
@competition = Competition.new(competition_params)
```

pingskills commented:
Updated to use CLONEABLE_ATTRIBUTES in build_clone method

jonatanklosko commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -239,16 +232,25 @@ def edit
   def time_until_competition
-    @competition = Competition.find(params[:id])
+    @competition = Competition.new
     @competition.assign_attributes(competition_params)
```
```ruby
@competition = Competition.new(competition_params)
```

jonatanklosko commented:
Ok, last two comments for you and that LGTM!

pingskills commented:
OK. I've made those changes :)

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -15,18 +15,23 @@ class Competition < ActiveRecord::Base
   has_many :media, class_name: "CompetitionMedium", foreign_key: "competitionId", dependent: :delete_all
 
+  CLONEABLE_ATTRIBUTES = %w(cityName countryId information eventSpecs venue venueAddress venueDetails website latitude longitude contact remarks use_wca_registration guests_enabled).freeze
+  UNCLONEABLE_ATTRIBUTES = %w(id name year month day endMonth endDay cellName showAtAll isConfirmed registration_open registration_close results_posted_at).freeze
```
This looks really awesome now, Jeff!

Could you split these 2 lines into one line per attribute? That means that future diffs will be easier to read, and the chances of conflicts are smaller. As we work on #329, I suspect that Competitions will continue to gather more and more attributes.

pingskills replied:
Sure that makes a lot of sense. I'll split those lines out.

jfly commented:
One comment from me about splitting the `CLONEABLE_ATTRIBUTES` and `UNCLONEABLE_ATTRIBUTES` into multiple lines, but otherwise LGTM!

pingskills commented:
I separated out CLONEABLE_ATTRIBUTES and UNCLONEABLE_ATTRIBUTES onto a separate line for each attribute. I wasn't really sure the best way to format that so let me know if you think it's OK how I did it. Thanks.

jfly commented on WcaOnRails/app/models/competition.rb:
```diff
@@ -15,18 +15,48 @@ class Competition < ActiveRecord::Base
+                            contact
+                            remarks
+                            use_wca_registration
+                            guests_enabled).freeze
```
This looks great! I might have done this to reduce indentation and not make the first/last ones special:

```ruby
CLONEABLE_ATTRIBUTES = %w(
  cityName
  countryId
  information
  eventSpecs
  venue
  venueAddress
  venueDetails
  website
  latitude
  longitude
  contact
  remarks
  use_wca_registration
  guests_enabled
).freeze
```

jonatanklosko replied:
What about grouping related fields? This would save some space, but I'm not sure what's better =)
```ruby
CLONEABLE_ATTRIBUTES = %w(
  countryId cityName
  venue venueAddress venueDetails
  contact website information remarks
  eventSpecs
  latitude longitude
  use_wca_registration
  guests_enabled
).freeze
UNCLONEABLE_ATTRIBUTES = %w(
  id name cellName
  year month day
  endMonth endDay
  showAtAll isConfirmed
  registration_open registration_close
  results_posted_at
).freeze

```

jfly commented:
LGTM! :ship: :shipit:
